### ISSUE TITLE
7-beta: Shadowing addon files doesn't always work

### ISSUE THREAD

**Issue by buschtoens** [babel/ember-cli-babel#240]
Cross-posting from #topic-typescript, so it doesn't get lost:

I noticed a peculiar problem with the https://github.com/babel/ember-cli-babel/pull/140. I’m shadowing files provided by addons. For example, `ember-intl` re-exports the `intl` service and `t` helper in their `app` directory. I’m shadowing these with my own compatibility layer `intl` service and `t` helper. I’m doing the exact same thing to the `features` service `ember-feature-flags`. This used to work reliably with `ember-cli-babel` 6.

But now it seems that _sometimes_ this shadowing does not work and the addon-provided files “leak through”. I’m also unable to find my override file in the Chrome Dev Tools. Triggering a re-build on the “missing” file by `touch`ing it fixes the problem.

Does it maybe have something to do with parallelization?
Concretely what happened is that my `t` helper was sometimes missing and the `ember-intl` one leaked through.

Same happened to the `features` service.

One side-note: I’m using my https://github.com/babel/ember-cli-babel/pull/236 branch to process `.ts` files.

This might be part of the cause.

**Comment by buschtoens** [comment]
Okay, so I am now 💯 certain, that this is caused by having overrides with a `.ts` extension, as enabled by #236. Since I made the hotfix of renaming all overrides to `.js`, I did not experience any flakiness.

As far as I understand the plan with ember-cli-typescript is to move into the direction of using ember-cli-babel to do the compilation, so I guess this is an issue we need to fix?

**Comment by dfreeman** [comment]
Yep, addons' `app` trees are merged with (and transpiled alongside) the host's, with the host's files taking precedence when there's a naming conflict. Unfortunately, with multiple possible extensions, two different input paths can now target the same output, and by that point in the process we don't have any way of knowing which file came from which source.

With MU this issue goes away (since there's no longer any merging like that), but in the interim it's likely to be annoying. I _could_ imagine updating `broccoli-babel-transpiler` to give preference based on the order of the configured extensions, but that seems kind of footgun-y (not to mention tricky with parallelized builds).
